### PR TITLE
Log loaded extensions when `PUMA_DEBUG` is set

### LINKED ISSUE #3020: Log loaded extensions behind `PUMA_DEBUG`
Re the 'loaded extensions', maybe something like the below?

Single mode:
```
* Listening on http://127.0.0.1:40001
* Loaded Extensions:
*     /mnt/c/GitHub/puma/vendor/bundle/ubuntu22/ruby/3.2.0+1/gems/nio4r-2.5.8/lib/nio4r_ext.so
*     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/date_core.so
*     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/enc/encdb.so
*     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/enc/trans/transdb.so
*     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/etc.so
*     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/io/wait.so
*     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/monitor.so
*     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/pathname.so
*     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/socket.so
*     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/stringio.so
Use Ctrl-C to stop
```

Cluster mode:
```
[16449] * Listening on http://127.0.0.1:40001
[16449] * Loaded Extensions:
[16449] *     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/date_core.so
[16449] *     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/enc/encdb.so
[16449] *     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/enc/trans/transdb.so
[16449] *     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/etc.so
[16449] *     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/io/wait.so
[16449] *     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/monitor.so
[16449] *     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/pathname.so
[16449] *     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/socket.so
[16449] *     /usr/local/lib/ruby/3.2.0+1/x86_64-linux/stringio.so
[16449] Use Ctrl-C to stop
```

**EDIT:** Above were generated without io-wait in Gemfile.lock

_Originally posted by @MSP-Greg in https://github.com/puma/puma/issues/2843#issuecomment-1148848255_

### PR DESCRIPTION
**Pull request by MSP-Greg**
### Description

This implements the feature requested in https://github.com/puma/puma/issues/3020, printing the loaded extensions if PUMA_DEBUG is set.

When Puma is running 'single' (no workers), the log is done in the main Puma process, after `Server#run` has been called, as some extensions maybe loaded in that call. For Puma running 'clustered' (workers used), the log is generated in Worker index 0, after `Server#run` is called and also in the master process.

Log is also generated on phased-restarts.

Added a test for both, small refactor (moved functionality to `intergration.rb`) for using a control server in CI.

Closes #3020

### Your checklist for this pull request
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by dentarg** [comment]
Re: tests, I see there's [`wait_for_server_to_include`](https://github.com/puma/puma/blob/7b6301009321d527891e4761da0d39a025b9015f/test/helpers/integration.rb#L129), maybe that could be used? Just to look for the `Loaded Extensions` string at least (when starting with `PUMA_DEBUG` set)

**Comment by MSP-Greg** [comment]
Added tests.  After #3035, I may try to add a count of the extensions to the two tests, so we'll know if there are additions.

Question - while running locally, I wondered if Puma in 'cluster' mode should log **both** master process and worker0 loaded extensions?  Currently it only logs worker0.

**Comment by dentarg** [comment]
> Added tests. After #3035, I may try to add a count of the extensions to the two tests, so we'll know if there are additions.

👍 
 
> Question - while running locally, I wondered if Puma in 'cluster' mode should log **both** master process and worker0 loaded extensions? Currently it only logs worker0.

I can only see it as a positive to log both

File: lib/puma/cluster/worker.rb [hunk 1]
```diff
@@ -115,6 +115,11 @@ def run
 
         while restart_server.pop
           server_thread = server.run
+
+          if @log_writer.debug? && index == 0
+            debug_loaded_extensions "────────────────────────────────── Loaded Extensions - worker 0:"
+          end
```
Comment (dentarg) [inline lib/puma/cluster/worker.rb:121]:
```suggestion
          debug_loaded_extensions("────────────────────────────────── Loaded Extensions - worker 0:") if index == 0
```

**Review by dentarg — COMMENTED** [review]
`debug_loaded_extensions` includes `if @log_writer.debug?` so no need to guard the method calls with it, right?

**Comment by MSP-Greg** [comment]
@dentarg

> `debug_loaded_extensions` includes `if @log_writer.debug?` so no need to guard the method calls with it, right?

Not sure.  I felt it was unlikely that many people would regularly be using `PUMA_DEBUG`, so guard the method call so it's never performed.  The guard in the method could be removed, but once it's been called, `LogWriter#debug` has the conditional anyway, so adding one more seemed insignificant?

**Comment by dentarg** [comment]
If we want to remove `if @log_writer.debug?` from the `debug_loaded_extensions` definition, we should at least be consistent and guard the call in `puma/lib/puma/single.rb` too, no?

https://github.com/puma/puma/blob/01b013a00da386fd8db66f6ad42460e7affc9cf1/lib/puma/single.rb#L60

**Comment by MSP-Greg** [comment]
> we should at least be consistent and guard the call in puma/lib/puma/single.rb too, no?

Sorry and thanks for pointing that out, my mistake.  Fixed in single.rb, removed conditional in runner.rb, with a comment...

File: lib/puma/runner.rb [hunk 2]
```diff
@@ -194,5 +194,12 @@ def stats
         }
       }
     end
+
+    # this method call should always be guarded by `@log_writer.debug?`
+    def debug_loaded_extensions(str)
```
Comment (dentarg) [inline lib/puma/runner.rb:199]: Why not have the guard in this method?

Also, I think "Puma running in debug mode" (`PUMA_DEBUG` set) is not something only `LogWriter` should know about. In the future, more things could be happening in "debug more" (e.g. collecting more details stats) in my mind.

What do you think of adding a constant `PUMA_DEBUG` to `lib/puma.rb`? And a method `Puma.debug?` (or `Puma.debug_mode?`) and use that here. That method can also be used in both `LogWriter` and `ErrorLogger` (both do `ENV.key?('PUMA_DEBUG')` today)

File: lib/puma/cluster/worker.rb [hunk 3]
```diff
@@ -115,6 +115,11 @@ def run
 
         while restart_server.pop
           server_thread = server.run
+
+          if @log_writer.debug? && index == 0
+            debug_loaded_extensions "────────────────────────────────── Loaded Extensions - worker 0:"
```
Comment (nateberkopec) [inline lib/puma/cluster/worker.rb:120]: I'm not a huge fan of having this long formatted string right in the code

Maybe:

```ruby
debug_loaded_extensions("worker 0")
```

...and let `debug_loaded_extensions` add the `---` padding and such.

File: lib/puma/runner.rb (hunk 2 above)
Comment (nateberkopec) [reply lib/puma/runner.rb:199]: I'm fine with either option, up to Greg.

**Review by nateberkopec — CHANGES_REQUESTED** [review]

File: lib/puma/cluster/worker.rb (hunk 3 above)
Comment (MSP-Greg) [reply lib/puma/cluster/worker.rb]: See [Fix !DRY code](https://github.com/puma/puma/pull/3036/commits/8d2da80a64587dbcf6f3d78a9ee309b974bd16e5)

Not my best code...  Sory for the double push, first was the commit, second was "maybe it's time to rebase" force push

Happy New Year!

**Review by nateberkopec — APPROVED** [review]

**Comment by nateberkopec** [comment]
Wait on merge until I push 6.0.2 though 😄

**Comment by nateberkopec** [comment]
@MSP-Greg Merge when ready 👍

**Review by dentarg — APPROVED** [review]

**Pull request merged**